### ISSUE TITLE
Undefined behaviour in FireSaber-KEM

### ISSUE THREAD

**Issue by claucece** [open-quantum-safe/liboqs#807]
Hi!

I've been looking at the library and running some things, and this is a really nice work, so thanks for all!

I recently ran some tests, enabling clang's UndefinedBehaviorSanitizer using the build sequence:

```
mkdir build && cd build
cmake -DCMAKE_BUILD_TYPE=Debug -DUSE_SANITIZER=Undefined -GNinja ..
ninja
```

My compiler is clang-11, and arch is x86_64-apple-darwin18.7.0

I ran onto several undefined behaviour, on some algorithms (I'm just pasting only one of each 'family'):

```
tests/test_kem FireSaber-KEM    
Configuration info
==================
Target platform:  x86_64-Darwin-18.7.0
Compiler:         clang (11.0.0 (clang-1100.0.33.17))
OQS version:      0.4.0-rc1
Git commit:       477ef3f60cd04f9b72b499d000a8f77ae99c0d89 (+ local modifications)
OpenSSL enabled:  Yes (OpenSSL 1.1.1g  21 Apr 2020)
AES:              OpenSSL
SHA-2:            OpenSSL
SHA-3:            OpenSSL

================================================================================
Sample computation for KEM FireSaber-KEM
================================================================================
../src/kem/saber/pqclean_firesaber_clean/poly_mul.c:50:44: runtime error: signed integer overflow: 57354 * 40101 cannot be represented in type 'int'
shared secrets are equal
```

```
tests/test_kem FrodoKEM-640-SHAKE
Configuration info
==================
Target platform:  x86_64-Darwin-18.7.0
Compiler:         clang (11.0.0 (clang-1100.0.33.17))
OQS version:      0.4.0-rc1
Git commit:       477ef3f60cd04f9b72b499d000a8f77ae99c0d89 (+ local modifications)
OpenSSL enabled:  Yes (OpenSSL 1.1.1g  21 Apr 2020)
AES:              OpenSSL
SHA-2:            OpenSSL
SHA-3:            OpenSSL

================================================================================
Sample computation for KEM FrodoKEM-640-SHAKE
================================================================================
../src/kem/frodokem/external/frodo_macrify_optimized.c:281:34: runtime error: signed integer overflow: 65534 * 59141 cannot be represented in type 'int'
../src/kem/frodokem/external/frodo_macrify_optimized.c:281:28: runtime error: signed integer overflow: 39658 + 2147450877 cannot be represented in type 'int'
shared secrets are equal
```

```
tests/test_kem SIDH-p503-compressed
Configuration info
==================
Target platform:  x86_64-Darwin-18.7.0
Compiler:         clang (11.0.0 (clang-1100.0.33.17))
OQS version:      0.4.0-rc1
Git commit:       477ef3f60cd04f9b72b499d000a8f77ae99c0d89 (+ local modifications)
OpenSSL enabled:  Yes (OpenSSL 1.1.1g  21 Apr 2020)
AES:              OpenSSL
SHA-2:            OpenSSL
SHA-3:            OpenSSL

================================================================================
Sample computation for KEM SIDH-p503-compressed
================================================================================
../src/kem/sike/external/P503/../compression/dlog.c:108:35: runtime error: index 4 out of bounds for type 'const f2elm_t' (aka 'felm_t const[2]')
../src/kem/sike/external/P503/../compression/dlog.c:96:39: runtime error: index 44 out of bounds for type 'const f2elm_t' (aka 'felm_t const[2]')
../src/kem/sike/external/P503/../compression/dlog.c:98:39: runtime error: index 44 out of bounds for type 'const f2elm_t' (aka 'felm_t const[2]')
../src/kem/sike/external/P503/../compression/dlog.c:115:35: runtime error: index 4 out of bounds for type 'const f2elm_t' (aka 'felm_t const[2]')
../src/kem/sike/external/P503/../compression/dlog.c:65:34: runtime error: index 4 out of bounds for type 'const f2elm_t' (aka 'felm_t const[2]')
../src/kem/sike/external/P503/../compression/dlog.c:58:38: runtime error: index 22 out of bounds for type 'const f2elm_t' (aka 'felm_t const[2]')
shared secrets are equal
../src/kem/sike/external/P503/../compression/sidh_compressed.c:1031:26: runtime error: index 233 out of bounds for type 'const uint64_t [17][8]'
```

I haven't finish running it all for all tests; but if it is of interest, I can do that.

**Comment by dstebila** [comment]
Hi Sofía! We are aware of some undefined behaviour (#637, #791) but not all of the ones you found, perhaps due to a different version of clang being used.  

I've tried fixing the Frodo ones, please see https://github.com/open-quantum-safe/liboqs/tree/ds-frodo-undefined-behaviour.

For SIDH, those should be covered by #637.

For Saber, can you file the bug report in PQClean, as we pull our Saber implementation from there?

**Comment by claucece** [comment]
> We are aware of some undefined behaviour (#637, #791) but not all of the ones you found, perhaps due to a different version of clang being used.

Yes, I think that might be the cause. There was lots of new things added to Clang 10.0: https://releases.llvm.org/10.0.0/tools/clang/docs/ReleaseNotes.html . I'll check if it is possible to add a machine on the CI with clang 11. Will that be of interest?

> I've tried fixing the Frodo ones, please see https://github.com/open-quantum-safe/liboqs/tree/ds-frodo-undefined-behaviour.

It is fixed ;)

> For Saber, can you file the bug report in PQClean, as we pull our Saber implementation from there?

For sure! I'll open a bug report over there.

**Comment by claucece** [comment]
Opened the issue over there: https://github.com/PQClean/PQClean/issues/309

Should we close this one then?

**Comment by dstebila** [comment]
> Opened the issue over there: [PQClean/PQClean#309](https://github.com/PQClean/PQClean/issues/309)
> 
> Should we close this one then?

I'll keep this open but with a new title for just FireSaber so that we remember to pull that in when it's fixed.

**Comment by dstebila** [comment]
Closing.  This was fixed by #827, and FireSaber passes when I run undefined behaviour sanitizer on current liboqs.
